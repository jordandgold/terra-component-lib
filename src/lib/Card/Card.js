import React, { Component } from "react";
import "./Card.scss";
import Button from "../Button/Button";
import ButtonLink from "../ButtonLink/ButtonLink";

import PropTypes from "prop-types";

class Card extends Component {
  render() {
    const { image, title, text, button, buttonLink, link } = this.props;

    return (
      <article className="ter-card">
        {image && (
          <figure className="ter-card__image">
            <img src={image.url} alt={image.altText} />
          </figure>
        )}
        <div className="ter-card__body">
          <h4 className="ter-card__title">{title}</h4>
          {text && <p className="ter-card__text">{text}</p>}

          {button && (
            <Button
              name={button.name}
              text={button.text}
              className="ter-button--primary--1"
              onClick={button.onClick}
            />
          )}
          {buttonLink && (
            <ButtonLink
              link={buttonLink.link}
              text={buttonLink.text}
              className="ter-button--primary--1"
            />
          )}
          {link && (
            <a className="ter-card__link" href={link.link}>
              {link.text}
            </a>
          )}
        </div>
      </article>
    );
  }
}

export default Card;

Card.propTypes = {
  image: PropTypes.object,
  title: PropTypes.string.isRequired,
  text: PropTypes.string,
  button: PropTypes.shape({
    onClick: PropTypes.func.isRequired,
    text: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired
  }),
  buttonLink: PropTypes.shape({
    link: PropTypes.string.isRequired,
    text: PropTypes.string.isRequired
  }),
  link: PropTypes.object
};
