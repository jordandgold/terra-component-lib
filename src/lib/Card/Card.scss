@import "../global-styles/terra.scss";

.ter-card {
  background: $card-bg;
  position: relative;
  border-width: $card-border-width;
  border-style: solid;
  border-color: $card-border-color;
  border-radius: $card-border-radius;
  display: flex;
  flex-direction: column;
  box-shadow: $card-box-shadow;
  &[href] {
    display: block;
  }
  &__body {
    font-size: $card-font-size;
    display: block;
    padding: spacing("md") spacing("sm");
    display: flex;
    flex: 1;
    flex-direction: column;
    p {
      font-size: $card-font-size;
    }
  }
  &__image {
    margin: spacing("none");
    padding: spacing("none");

    img {
      width: 100%;
      height: auto;
    }
  }
}
