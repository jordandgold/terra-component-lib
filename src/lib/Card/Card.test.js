import React from "react";
import { shallow } from "enzyme";

import Card from "./Card";

describe("Card", () => {
  let wrapper;

  const mockImage = {
    link: "http://www.mocklink.com/mockImage.jpg",
    description: "This is a mock image description"
  };
  const mockTitle = "This is a mock title";
  const mockText = [
    "This is the first paragraph",
    "This is the second paragraph"
  ];
  const mockButton = {
    name: "mock name",
    text: "mock text",
    className: "hello",
    onClick: jest.fn()
  };
  const mockButtonLink = {
    link: "mock link",
    text: "mock text"
  };
  const mockLink = {
    link: "http://www.mocklink.com",
    label: "I am a link!"
  };
  const mockCategory = "Hello! I am a category!";

  beforeEach(() => {
    wrapper = shallow(
      <Card
        image={mockImage}
        title={mockTitle}
        text={mockText}
        button={mockButton}
        buttonLink={mockButtonLink}
        link={mockLink}
        category={mockCategory}
      />
    );
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe("generateText", () => {
    it("should return some jsx", () => {
      const expected = [
        '<p class="card-text">This is the first paragraph</p>',
        '<p class="card-text">This is the second paragraph</p>'
      ];
      const result = wrapper
        .instance()
        .generateText()
        .map(point => {
          return shallow(point).html();
        });

      expect(result).toEqual(expected);
    });
  });
});
