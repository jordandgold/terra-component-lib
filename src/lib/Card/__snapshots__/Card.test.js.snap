// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match the snapshot 1`] = `
<article
  className="ter-card"
>
  <figure
    className="ter-card__image"
  >
    <img
      alt="This is a mock image description"
      src="http://www.mocklink.com/mockImage.jpg"
    />
  </figure>
  <div
    className="ter-card__body"
  >
    <h4
      className="ter-card__title"
    >
      This is a mock title
    </h4>
    <p
      className="ter-card__text"
    >
      This is the first paragraph
    </p>
    <a
      className="ter-card__link"
      href="http://www.mocklink.com"
    />
    <Button
      className="ter-button--primary--1"
      name="mock name"
      onClick={[MockFunction]}
      text="mock text"
    />
    <ButtonLink
      className="ter-button--primary--1"
      link="mock link"
      text="mock text"
    />
  </div>
</article>
`;
